Formal Verification

gsvarovsky opened the pull request:
Resolves #4 

See the [**write-up**](https://github.com/m-ld/m-ld-security-spec/blob/formal-verification/verification/formal%20verification.md) (`verification/formal verification.md`)

Angus-McAllister reviewed (approved):
Great piece of work!  Not being a practioner of CSP, I have only reviewed one of the `.csp` files.  Also, time constraints precluded a thorough review of `formal verification.md`, but if the track record of prior contributions is anything to go by, I have no doubt it is excellent.

gsvarovsky commented on design/security design.mm:
```diff
@@ -592,7 +592,7 @@
 </node>
 </node>
-<node TEXT="after agreement, no message can&#xa;be accepted which is not caused by it" ID="ID_317373433" CREATED="1628086333506" MODIFIED="1628667782196">
+<node TEXT="after agreement, no message can&#xa;be accepted which is not caused by it" FOLDED="true" ID="ID_317373433" CREATED="1628086333506" MODIFIED="1628667782196">
```
Note added in the mindmap.

Angus-McAllister commented:
Having now taken some time to review [`formal verification.md`](https://github.com/m-ld/m-ld-security-spec/blob/main/verification/formal%20verification.md), I have discovered some shortcomings in it.  Whilst some of these are minor grammatical ones, a more concerning issue is that of references to prior art.  The bibliography links seem not to have survived; only the one-word keys appear in the document (e.g. \[shapiroConflictFreeReplicatedData2011\] in [related work](https://github.com/m-ld/m-ld-security-spec/blob/main/verification/formal%20verification.md#related-work).

One clarity-related observation is that for the [convergent state](https://github.com/m-ld/m-ld-security-spec/blob/main/verification/formal%20verification.md#convergent-state) section, the Strong Eventual Consistency requirements are confusingly expressed, particularly in the inconsistent relationships between the terms update, replica and correct.  In 1., an update is delivered at a replica, yet in 2., it is the replicas that perform the delivery of updates, an apparent contradiction.

Also in the realm of clarity, the same section describes CSP 'pipes' as "processes accepting a `send` event specific *to some source processes*".  The wording is ambiguous, in that it isn't clear whether the `send` event is going to those source processes, or is instead specific to those source processes.  Suggested re-wording is to mention the origin of those messages, and to make it clear that it is the event carrying the data payload, which would help clear up the ambiguity.

Further to this section, another source of confusion is the reference to `receive` events as being sent.  Perhaps using a different label for such events might avoid that confusion arising.

Finally for this section, the diagram does not make it clear whether the items labelled `RemoteTransmit(x,y)` are messages, events, or merely expressions of the relationships between different clones.

gsvarovsky commented:
> The bibliography links seem not to have survived

These are textual references only, to entries in [references.bib](https://github.com/m-ld/m-ld-security-spec/blob/main/references.bib); this is the simple system that has been used throughout this project.

> the Strong Eventual Consistency requirements are confusingly expressed

These are verbatim from the Shapiro paper. In the definition of SEC, it doesn't matter how delivery happens.

> The wording is ambiguous, in that it isn't clear whether the `send` event is going to those source processes, or is instead specific to those source processes.

The confusion here is because in CSP an event is symmetrical with respect to the processes being modelled; it's just an abstract synchronisation point. So 'Accept' is unfortunately overloaded – in CSP, and here, it means to participate in an event. I have switched to the latter term and added a reference to the CSP book section.

> another source of confusion is the reference to `receive` events as being sent. Perhaps using a different label for such events might avoid that confusion arising

I can't find that any more. The `send` and `receive` labels form part of the CSP standard pattern for communication, so I don't want to change them (although amusingly, the book referenced uses different labels).

> the diagram does not make it clear whether the items labelled `RemoteTransmit(x,y)` are messages, events, or merely expressions of the relationships between different clones

They are CSP processes, each corresponding to an abstract 'pipe', and they are named in the `.csp` file. I have clarified. Note also the comment about the apparent mesh network: a 'process' in CSP is abstract and may not correspond to any recognisable implementation artefact.
